braidpool_spec.md

```diff
@@ -460,33 +675,73 @@ formed, fully signed and valid bitcoin transactions that can be broadcast. See
 [Payout Authorization](#payout-authorization) for considerations on how to
 sign/authorize this UHPO transaction.
 
+[kp]
+
+The UHPO transactions are signed and broadcast using the braidpool P2P
+network. In case of the pool shutting down, the miners will have to figure out a
+communication channel to broadcast their signature shares and reach agreement on
+spending the UHPO.
+
+This sounds like a problem.
+
+[kp]
+
 We don't ever want to actually have to broadcast this UHPO set transaction
 except in the case of pool failure. Similar to other optimistic protocols like
 Lightning, we will withhold this transaction from bitcoin and update it
-out-of-band with respect to bitcoin. With each new block we will update the UHPO
-set transaction to account for any new shares since the last block mined by the
-pool.
+out-of-band with respect to bitcoin. With each new block found by braidpool, we
+will update the UHPO set transaction to account for any new shares since the
+last block mined by the pool.
 
 Furthermore a decentralized mining pool should support "withdrawal" by hashers.
-This would take the form of a special message or transaction sent to the pool
+This will take the form of a special message or transaction sent to the pool
 (and agreed by consensus within the pool) to *remove* a hasher's output from the
 UHPO set transaction, and create a new separate transaction which pays that
 hasher, [authorizes](#payout-authorization) it, and broadcasts it to bitcoin.
 Miners may not cash out for the *current* difficulty adjustment window (because
 the share/BTC price is not yet decided), but may only cash out for the *last*
 (and older) difficulty adjustment window(s).
 
+[kp]
+
+We need to address potential race condition and instability periods when miners
+are cashing out and new bitcoin blocks are being found. If there is no such
+instability, we should explain why. I expect this question will be raised during
+review.
+
+
+[/kp]
+
+
 The share value for the *current* difficulty adjustment epoch is encoded
 proportionally in the UHPO transactions, however this is only for use in the
 case of catastrophic failure of Braidpool. During normal operation, the UHPO
 transaction is fixed at the end of the difficulty adjustment window when the
 share/BTC price for that epoch is known.
 
+[kp]
+
+Why is the share value fixed?
+
+We don't need a fixed share value. A miner's share value over a period can
+determined by their claimed difficulty. The difficulty is claimed at the outset,
+then all shares confirm the miner mined with that difficulty as some shares will
+get close to that difficulty.
+
+[/kp]
+
```

Comment:
The share value is determined by both (1) difficulty and (2) earned income (number of blocks won and fees).

After (2) is determined at the end of the epoch, the share value can be calculated and indeed all beads with that difficulty will have the same value, up to latency punishment.

The point of the above paragraph is that the UHPO transaction is constructed *before* the end of the epoch and before (2) is known. Therefore a different calculation must be used. But it's not that big a deal, you just pretend that the current epoch ends at the current block.